packages/angular_devkit/build_angular/src/builders/browser-esbuild/schema.json

```diff
@@ -14,7 +14,7 @@
     },
     "main": {
       "type": "string",
-      "description": "The full path for the main entry point to the app, relative to the current workspace."
+      "description": "The full path for the main entry point to the app. May be a relative or absolute path. If given a relative path, it is resolved relative to the current workspace."
```

Comment:
The motivation here is to support injecting code by the build process. In particular for testing, we need to inject `TestBed.initTestEnvironment()` which will not be within the user's workspace. I agree in general, users should not bundle code from outside the workspace, but this is one case where we effectively need to inject a file from `node_modules/...` which may not be in the workspace (ex. Yarn PnP).

I could probably limit this to `entryPoints` which would be restricted to programmatic invocations of ESBuild-er, that way `main` could not be abused in this manner and hopefully limit "bad" usages in the wild, though it would also be confusing that `main` and `entryPoints` resolve files differently. In that case, I would probably want to ban absolute paths altogether from `main` instead of treating them like relative paths.

Reply:
> The motivation here is to support injecting code by the build process. In particular for testing, we need to inject TestBed.initTestEnvironment() which will not be within the user's workspace. I agree in general, users should not bundle code from outside the workspace, but this is one case where we effectively need to inject a file from node_modules/... which may not be in the workspace (ex. Yarn PnP).

I think for the time being as a temporary solution, but we definitly need to better approach for the in built `test.ts`. While we could get away with bundling a file in `node_module` I have doubts that it will solve Yarn PNP causes as this could cause module resolutions errors, because esbuild will try to resolve `@angular/` from in `@angular-devkit/build-angular` which is incorrect as it should be resolved from the users workspace root. 

Eventually, we should explore virtual modules or something like we currently have for [webpack](https://github.com/angular/angular-cli/blob/866263b21e71d03800ae91033e951f9407ad0547/packages/angular_devkit/build_angular/src/builders/karma/index.ts#L242-L260), as something similar as this will also be needed for SSR in `16.1`.

> I would probably want to ban absolute paths altogether from main instead of treating them like relative paths.

I agree with this, but in this case, I think we need to do this for all builders options as otherwise it will be confused that in some cases we treating absolute like relative paths and sometimes not. Also, this sounds like a thing we'd want to do in a major with a migration due to the breaking nature of it.

Reply:
Agreed that virtual modules would make this better and there is a potential package resolution issue with `pnpm`. This should be good enough for now and we can update it to use virtual modules once we have a better idea of what that would look like.

For now, I limited absolute paths to just the new internal `entryPoints` option. `main` is mostly unaffected (only new behavior is that it throws when an empty string is passed in). That should limit the blast radius of this particular change until we can leverage virtual modules here.